crates/sound/src/device.rs

```diff
@@ -388,100 +388,76 @@ impl VhostUserSoundThread {
-                    {
-                        return Err(Error::UnexpectedDescriptorSize(
-                            u32::from(
-                                self.streams.read().unwrap()[stream_id as usize]
-                                    .params
-                                    .buffer_bytes,
-                            ) as usize,
-                            descriptor.len(),
-                        )
-                        .into());
-                    }
-                    TxState::WaitingBufferForStreamId(_stream_id) => {
-                        /*
-                        Rather than copying the content of a descriptor, buffer keeps a pointer to it.
-                        When we copy just after the request is enqueued, the guest's userspace may or
-                        may not have updated the buffer contents. Guest driver simply moves buffers
-                        from the used ring to the available ring without knowing whether the content
-                        has been updated. The device only reads the buffer from guest memory when the
-                        audio engine requires it, which is about after a period thus ensuring that the
-                        buffer is up-to-date.
-                        */
-                        buffers.push(Buffer::new(*descriptor, Arc::clone(&message)));
-                    }
+            let desc_request = descriptors[0];
+
+            if desc_request.len() as usize != size_of::<VirtioSoundPcmXfer>() {
+                return Err(Error::UnexpectedDescriptorSize(
+                    size_of::<VirtioSoundPcmXfer>(),
+                    desc_request.len(),
+                )
+                .into());
+            }
+
+            if desc_request.is_write_only() {
+                return Err(Error::DescriptorReadFailed.into());
+            }
+
+            let hdr_request = desc_chain
+                .memory()
+                .read_obj::<VirtioSoundPcmXfer>(desc_request.addr())
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            let stream_id = hdr_request.stream_id.to_native();
+
+            // TODO: to check if stream_id is in correct state
+            stream_ids.insert(stream_id);
+
+            let mut len = 0;
+
+            let data_descs = &descriptors[1..descriptors.len() - 1];
+
+            for data in data_descs {
+                if data.is_write_only() {
+                    return Err(Error::DescriptorReadFailed.into());
                 }
+                len += data.len()
+            }
+
+            let status_desc = descriptors[descriptors.len() - 1];
+            if status_desc.len() as usize != size_of::<VirtioSoundPcmStatus>() {
+                return Err(Error::UnexpectedDescriptorSize(
+                    size_of::<VirtioSoundPcmStatus>(),
+                    status_desc.len(),
+                )
+                .into());
+            }
```

Comment:
Code just moved around, but: The spec probably allows all these structs to be broken up across multiple descriptors?